src/QsCompiler/Transformations/CallGraphWalker.cs

```diff
@@ -0,0 +1,344 @@
+            {
+                if (!finished.Contains(key))
+                {
+                    processDependencies(key);
+                }
+            }
+
+            return cycles;
+        }
+    }
+
+    /// <summary>
+    /// This transformation walks through the compilation without changing it, building up a call graph as it does.
+    /// This call graph is then returned to the user.
+    /// </summary>
+    public static class BuildCallGraph
+    {
+        public static CallGraph Apply(QsCompilation compilation)
+        {
+            var walker = new BuildGraph();
+
+            foreach (var ns in compilation.Namespaces)
+            {
+                walker.Namespaces.OnNamespace(ns);
+            }
+
+            return walker.SharedState.graph;
+        }
+
+        private class BuildGraph : SyntaxTreeTransformation<BuildGraph.TransformationState>
+        {
+            public class TransformationState
+            {
+                internal QsSpecialization spec;
+
+                internal bool inCall = false;
+                internal bool adjoint = false;
+                internal bool controlled = false;
+
+                internal CallGraph graph = new CallGraph();
+            }
+
+            public BuildGraph() : base(new TransformationState())
+            {
+                this.Namespaces = new NamespaceTransformation(this);
+                this.Statements = new StatementTransformation<TransformationState>(this, TransformationOptions.NoRebuild);
+                this.StatementKinds = new StatementKindTransformation<TransformationState>(this, TransformationOptions.NoRebuild);
+                this.Expressions = new ExpressionTransformation<TransformationState>(this, TransformationOptions.NoRebuild);
+                this.ExpressionKinds = new ExpressionKindTransformation(this);
+                this.Types = new TypeTransformation<TransformationState>(this, TransformationOptions.Disabled);
+            }
+
+            private class NamespaceTransformation : NamespaceTransformation<TransformationState>
+            {
+                public NamespaceTransformation(SyntaxTreeTransformation<TransformationState> parent) : base(parent, TransformationOptions.NoRebuild) { }
+
+                public override QsSpecialization OnSpecializationDeclaration(QsSpecialization spec)
+                {
+                    SharedState.spec = spec;
+                    return base.OnSpecializationDeclaration(spec);
+                }
+            }
+
+            private class ExpressionKindTransformation : ExpressionKindTransformation<TransformationState>
+            {
+                public ExpressionKindTransformation(SyntaxTreeTransformation<TransformationState> parent) : base(parent, TransformationOptions.NoRebuild) { }
+
+                private ExpressionKind HandleCall(TypedExpression method, TypedExpression arg)
+                {
+                    var contextInCall = SharedState.inCall;
```

Comment:
>var [](start = 20, length = 3)

Is it bool? Just one more character and readability is greatly improved.

Reply:
I would have preferred to use explicit types everywhere throughout the code, but the style for this code base is to use var when possible, despite the hit to readability.